iocore/nexthop/NextHopHost.h

```diff
@@ -0,0 +1,201 @@
+#include <atomic>
+#include "I_EventSystem.h"
+#include "PartitionedMap.h"
+#include "PropertyBlock.h"
+#include "ts/string_view.h"
+
+/**
+ * @file
+ *
+ * @brief this file defines structures to store data about each host and ip.
+ *
+ * HostAddr - stores a IpEndpoint, expire time, and up status. Additional properties can be defined at system start.
+ *
+ * HostName - stores a FQDN and a list of HostAddr. Additional properties can be defined at system start.
+ *  All references to HostNames will be shared pointers. They are only deleted when all references are deleted.
+ *
+ * Lookups are performed with one global mutex lock to protect integrity of the map.
+ * @see HostLookupByName, HostLookupByAddr
+ *
+ * When you are reading/writing a HostNamePtr, ensure that you find the assigned mutex from the LockPool. This ensures a single
+ * reader/writer to the HostName, associated HostAddrs and property blocks.
+ * @see HostName::getMutex()
+ *
+ */
+
+// Define a std::hash<>() for the lookmaps
+std_hasher_macro(IpEndpoint, ip, ats_ip_port_hash(ip));
+
+namespace NextHop
+{
+using Mutex = RWLock;
+
+/// uid for a host that should persist between updates and reloads.
+/// It is prefered for other systems to cache HostID instead of a HostNamePtr.
+using HostID = size_t;
+
+struct HostName;
+using HostNamePtr = std::shared_ptr<HostName>;
+
+/// lookup HostID, direct string hash
+HostID getHostID(ts::string_view hostname);
+
+/// lookup HostID with map
+HostID getHostID(IpEndpoint addr);
+
+/// lookup HostNamePtr
+HostNamePtr getHost(HostID);
+HostNamePtr getHost(IpEndpoint addr);
+HostNamePtr getHost(ts::string_view hostname);
+
+/// our internal lookup maps
+// TODO move to .cc
+LookupMap<HostID, HostNamePtr> HostLookupByNameHash;
+LookupMap<IpEndpoint, HostNamePtr> HostLookupByAddr;
+
+/// Defines a host address, status, and a block of associated properties.
+/** Only variables need by TSCore are in this structure.
+ * Plugins should define property blocks.
+ * @see PropBlockDeclare
+ */
+struct HostAddr : public PropertyBlock<HostAddr> {
+  const IpEndpoint m_addr;      ///< ip and port
+  std::atomic_uint m_eol;       ///< "end of life" time, set by DNS time to live
+  std::atomic_bool m_available; ///< true when this IP is available for use
+
+  HostAddr(const IpEndpoint &ip_addr, HostNamePtr host) : m_addr(ip_addr) { HostLookupByAddr.put(ip_addr, host); }
+
+  // Note, this uses PropertyBlock::new and delete to manage.
+
+private:
+  HostAddr();
```

Comment:
Why is this private?

Reply:
to force people to use HostAddr::alloc().